### PR TITLE
[SPARK-26014][R] Deprecate R prior to version 3.4 in SparkR

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to bump up the minimum versions of R from 3.1 to 3.4.

R version. 3.1.x is too old. It's released 4.5 years ago. R 3.4.0 is released 1.5 years ago. Considering the timing for Spark 3.0, deprecating lower versions, bumping up R to 3.4 might be reasonable option.

It should be good to deprecate and drop < R 3.4 support.

## How was this patch tested?

Jenkins tests.

### CODE REVIEW

**Comment by HyukjinKwon** [comment]
Tests probably will fail since it produces warnings. 

cc @felixcheung. @shaneknapp, @viirya, @shivaram, @falaki, @mengxr, @yanboliang FYI.

This PR is made per http://apache-spark-developers-list.1001551.n3.nabble.com/discuss-SparkR-CRAN-feasibility-check-server-problem-td25605.html

**Comment by HyukjinKwon** [comment]
adding @srowen too.

File: docs/index.md [hunk 1]
```diff
@@ -31,7 +31,8 @@ Spark runs on both Windows and UNIX-like systems (e.g. Linux, Mac OS). It's easy
 locally on one machine --- all you need is to have `java` installed on your system `PATH`,
 or the `JAVA_HOME` environment variable pointing to a Java installation.
 
-Spark runs on Java 8+, Python 2.7+/3.4+ and R 3.1+. For the Scala API, Spark {{site.SPARK_VERSION}}
+Spark runs on Java 8+, Python 2.7+/3.4+ and R 3.1+. R prior to version 3.4 is deprecated as of Spark 3.0.
```
Comment (viirya) [inline docs/index.md]: 3.1+ -> 3.4+?

Comment (HyukjinKwon) [reply docs/index.md]: Hmmmmm .. I was thinking we could change them when we actually drop the support. Technically it does support 3.1+ yet although 3.1, 3.2, and 3.3 are deprecated.

Comment (viirya) [reply docs/index.md]: hmm, so R prior to version 3.4 is just deprecated, not dropped in in Spark 3.0?

Comment (HyukjinKwon) [reply docs/index.md]: Ah,  yea, I switched this to deprecate it for now. I was a bit curious about that.

**Review by srowen — APPROVED** [review]
I'll put a note in the Docs Text field in the JIRA for release notes.

**Comment by felixcheung** [comment]
I think this should say unsupported (ie could still work) instead of deprecated

Also the compareVersion should check both major and minor ie 3.4.0

**Comment by felixcheung** [comment]
Also I think the warning should be in .First in general.R

**Comment by HyukjinKwon** [comment]
Yea will take a look to address. But about documenting unsupported, if we explicitly are going to say it's unsupported and dropped, for instance, we should remove the compatibility change (https://github.com/apache/spark/blob/master/R/pkg/src-native/string_hash_code.c) and I assume previous versions don't work. Deprecation step might be more concervative and consistent with dropping steps of other language APIs.

File: R/pkg/R/sparkR.R [hunk 2]
```diff
@@ -283,6 +283,10 @@ sparkR.session <- function(
   enableHiveSupport = TRUE,
   ...) {
 
+  if (utils::compareVersion(paste0(R.version$major, ".", R.version$minor), "3.4.0") == -1) {
+    warning("R prior to version 3.4 is deprecated as of Spark 3.0.")
+  }
```
Comment (HyukjinKwon) [inline R/pkg/R/sparkR.R]: @felixcheung, I updated the condition to check major version as well.

BTW, if we add it in .First in general.R, looks it's not going to print out the warnings in the SparkR shell. When we drop PySpark, we did in a similar place (https://github.com/apache/spark/pull/15733/files#diff-01e513a19a2a6f31ecaabd3dd6ac12d9R192).

File: R/WINDOWS.md [hunk 3]
```diff
@@ -3,7 +3,7 @@
 To build SparkR on Windows, the following steps are required
 
 1. Install R (>= 3.1) and [Rtools](http://cran.r-project.org/bin/windows/Rtools/). Make sure to
-include Rtools and R in `PATH`.
+include Rtools and R in `PATH`. Note that R prior to version 3.4 is deprecated as of Spark 3.0.
```
Comment (dongjoon-hyun) [inline R/WINDOWS.md]: I'm wondering if we can deprecate this in Spark 2.4.1?

Comment (dongjoon-hyun) [reply R/WINDOWS.md]: I see. Thanks, @felixcheung .

**Comment by felixcheung** [comment]
I think it's easier to say unsupported if we are not testing it in jenkins or appveyer. I don't know if we any coverage at release for older R version anyway, so it's better to unsupported then deprecate. 

but agree maybe the way to do this is deprecate without updating R in jenkins

**Comment by felixcheung** [comment]
Hey shane I don’t think we are saying to test multiple R version at all. In fact quite the opposite, just the new(er) version at some point in the future.

(We don’t have a better solution for packages though. There’s another PR for R arrow package for example)

File: docs/index.md (hunk 1 above)
Comment (HyukjinKwon) [reply docs/index.md]: At least Python version was noted here before.

**Comment by HyukjinKwon** [comment]
@shaneknapp, do you roughly know how difficult it is (and do you have some time shortly) to upgrade R from 3.1 to 3.4? I am asking this because I had some difficulties when I tried to manually upgrade from a certain low version to another non-latest version.

If it's expected to take a while, let's go deprecation step.
If that's expected to be less difficult, let's go saying unsupporting way.

Does this sound okay to you @felixcheung?

**Comment by shaneknapp** [comment]
TL;DR:  let's go w/deprecation.

still TL;DR:  if i never have to install or manage R again, i will be a happy person!

@HyukjinKwon upgrading R is easy.  getting the right mix of R and all of the associated packages working "as expected" is a nightmare.

the biggest problem i foresee is if we upgrade R (and all other packages) on the workers, every version of spark will be tested against this...  and there will be bugs, test failures, and other time consuming (and obtuse) problems to debug.  multiply this by every branch, and you can see the rabbit hole you've just entered.

for example, a month ago when i finally had time to dive back in to the ubuntu port, after finally figuring out how to install R+friends on ubuntu in an identical way to the centos workers, i STILL was finding problems w/lintr (see:  https://github.com/apache/spark/pull/22896).

anyways:  i'm more than happy to upgrade R and all the packages to something much more recent, but i will definitely appreciate some help in the game of test-failure whack-a-mole.

**Comment by felixcheung** [comment]
Yea there are some problem with some packages we depend on that are not installable from CRAN (eg too old) so it will be hard to a new version of R and new installation.

So to clarify, depreciation as is we still test on R 3.1?

**Comment by shaneknapp** [comment]
@felixcheung @HyukjinKwon 

yes:  deprecation in this case means we test against R-3.1.1

**Comment by HyukjinKwon** [comment]
Yup, will address the other comments and update the PR accordingly.

**Review by felixcheung — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

Thanks @felixcheung.
